staging/src/k8s.io/api/coordination/v1alpha1/types.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.31
+
+// LeaseCandidate defines a candidate for a lease object.
+// Candidates are created such that coordinated leader election will pick the best leader from the list of candidates.
+type LeaseCandidate struct {
+	metav1.TypeMeta `json:",inline"`
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec contains the specification of the Lease.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Spec LeaseCandidateSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// LeaseSpec is a specification of a Lease.
+type LeaseCandidateSpec struct {
+	// BinaryVersion is the binary version
+	BinaryVersion string `json:"binaryVersion,omitempty" protobuf:"bytes,5,opt,name=binaryVersion"`
+	// CompatibilityVersion is the compatibility version
+	CompatibilityVersion string `json:"compatibilityVersion,omitempty" protobuf:"bytes,6,opt,name=compatiblityVersion"`
+	// TargetLease is the name of the lease that the candidate can lead
+	TargetLease string `json:"targetLease,omitempty" protobuf:"bytes,7,opt,name=targetLease"`
+
+	// leaseDurationSeconds is a duration that candidates for a lease need
+	// to wait to force acquire it. This is measure against time of last
+	// observed renewTime.
+	// +optional
+	LeaseDurationSeconds *int32 `json:"leaseDurationSeconds,omitempty" protobuf:"varint,2,opt,name=leaseDurationSeconds"`
+	// renewTime is the time that the LeaseCandidate was last updated.
+	// Unlike Lease objects, candidates are not refreshed frequently.
+	// Any time a Lease needs to do leader election, an annotation is sent
```

Comment:
Who is actually responsible for renewing the lease - the component trying to be elected or the CLE controller?

Is the proposal that the CLE writes an annotation (on this candidate or on the lease) to indicate "stop renewing ASAP"?

Reply:
so this was something that was touched on but implementation details weren't fully explained in the KEP and I'll try to do that here and get feedback.

The main reason we're opting for this logic is to avoid having frequent LeaseCandidate renewals and adding additional QPS to the apiserver.

We have two objects, Lease and LeaseCandidate.

For LeaseCandidate
- LeaseCandidates are renewed every 5 mins
- It's possible a LeaseCandidate goes away during the 5 mins, we obviously do not want a 5 min turnover time in selecting a new leader. 
- Watches itself for a specific annotation to respond to (more details in election)

For Lease:
- If a LeaseCandidate is elected as leader, it is responsible for renewing the Lease. Every LeaseCandidate watches the Lease object to see if it has been elected.
- If lease expires, CLE is responsible for running an election and finding a leader
- If a new LeaseCandidate joins the pool, CLE is responsible for running an election and checking if the current leader is still the most optimal.

Election:
- During an election, CLE will find all LeaseCandidates for a corresponding Lease
- It's possible that some LeaseCandidates are unavailable because they're only renewed every 5 mins
- For all eligible LeaseCandidates, CLE will update them with an annotation "coordination.k8s.io/status" = "Pending"
- Election will pause until either
  - 5 seconds have passed
  - all LeaseCandidates for the lease update their "coordination.k8s.io/status" to "Ack"
- CLE will only select the best leader out of the LeaseCandidates that have "Ack"'d the status annotation.
- CLE updates the Lease either with a EndOfTerm flag or with the name of the new LeaseCandidate as the holderIdentity

/cc @jpbetz @sttts
@aojea: This explains some of the rationale around the "Pending"/"Ack"

Reply:
There's a whole lot of annotating going on, each of which is an API operation of its own, in service of avoiding API operations.

A typical lease is about 15s, right?

A typical election is 3 candidates, but could be more, right?

So every 15s:
* CLE writes to every candidate (3 ops)
* Every candidate ACKs (3 ops)
* CLE maybe writes end-of-term
* Winner writes the lease

That's at least 7 writes, maybe 8, and more watch events.  If, instead, each LeaseCandidate was refreshed every 7.5 seconds (Nyquist?) that would be the same 6+1 ops but way simpler?  Am I missing something?  It's not like annotations are less expensive than any other field.

As for how to signal "end your term now", you need some acknowledgement, too, right?  Worst case is the overthrown leader runs out the lease's duration?

Reply:
Lease and LeaseCandidate renewal and writes are separate operations on completely different intervals.
We do not reelect every 15 seconsd

Every 15s
- Lease is renewed by the holder (Candidate) (1 op)

There should only be a single write. All other operation are based on changes

- If Lease expires, the loop you described runs and we have like 7-8 writes
- If a new LeaseCandidate joins, we need to double check election and have the loop with 7-8 writes

The idea is that in steady state, leases shouldn't expire, new leasecandidates shouldn't be joining frequently, so it should amortize to only the renewal write operation.

Reply:
Ahh, I see.  Thanks for clarifying.

I do not love annotations for this.  What if the way we signal "we need prrof-of-life" is to set the `LeaseCandidate.renewTime` to 0.  If a component see its own renewTime hit 0, then it is expected to set the renewTime as quickly as possible.

Reply:
Next, how to signal "end your term now".

Do we need some acknowledgement from the lease holder that it has seen the request and stopped acting as leader?  What if it doesn't ACK?  Worst case is the overthrown leader runs out the lease's duration?

Reply:
ah renewTime again :D. From our discussion with @sttts, gc on LeaseCandidate is necessary and I was planning to use renewTime to perform gc and would prefer not using renewTime to set it. I'd also like to differentiate between a LeaseCandidate's proof-of-life ack UPDATE on the resource and a LeaseCandidate's fields may-be-changed UPDATE. (LeaseCandidate acks should not trigger a reelection attempt while LeaseCandidate updates should)

Going through the options, RenewTime is used by the gc controller and I'd prefer not touching it. CandidateDurationSeconds could be an option but we would have to reset it back to its original value after the ack which seems troublesome. What are your thoughts on yet another new field?

Reply:
> I'd also like to differentiate between a LeaseCandidate's proof-of-life ack UPDATE on the resource and a LeaseCandidate's fields may-be-changed UPDATE.

Is there some mechanism I don't know about which differentiates that?  The only way I know how to do it is to compare fields I care about.  API server sends the same message, either way.

The "we use it for GC" is a good reason not to overload it.  There are no other fields to use, and this isn't really appropriate for an annotation, so I guess new field is our path.  What if we did something like:

```
// Ping is a request for the owner of this LeaseCandidate to update the renewTime as soon
// as possible.  This can be used as "proof of life" of the candidate.  The requester should set
// this field to the current time.  When the component which owns this object sees this
// field with a value, it should by setting a new renewTime and un-setting ping.
Ping *metav1.MicroTime
```

This gives the CLE a sense of how long the request took.  If we ever need the same capability for `Lease`, we can use the same protocol.  

What think?

Reply:
> Is there some mechanism I don't know about which differentiates that? The only way I know how to do it is to compare fields I care about. API server sends the same message, either way.
No, but I was trying to achieve that through the annotations logic. The scenario is that updates to a LeaseCandidate could prompt a better candidate being selected so we trigger an election. However, acking is also done through updates and we want to avoid triggering reelection on acks, otherwise we have an endless loop of election triggers ack, ack triggers election, etc.

I'd like to be able to incorporate the 3 states all from a single update operation (UPDATE=ack,UPDATE=re-elect,UPDATE=please-ack) within the proposed fields. CLE watches the LeaseCandidate object, and the two cases (UPDATE=ack, UPDATE=please-ack) should NOT trigger a re-election. Alternative idea with Ping.

```
// Ping is the last time that a "proof of life" of the candidate was requested.
// When a component which owns this object see this field with a value later than the current renewTime, it responds by setting a new renewTime.
//
// The field should be unset if the LeaseCandidate updates its strategy or fields used for leader election.
// Updates to this object that do not unset Ping will not trigger a re-election of the corresponding LeaseName that this object references.
// Unsetting this field will directly trigger a re-election of the corresponding LeaseName that this object references.
Ping *metav1.Time
```

During an election, ack updates the renewTime so we should be able to just look at renewTime to verify an ack.
The set of eligible candidates is given by looking at `LeaseCandidate.renewTime >= electionStartTime` for each candidate.

The three states I described can be verified with:

- (UPDATE=ack): LeaseCandidate.renewTime > electionStartTime (only CLE can determine this)
- (UPDATE=re-elect): LeaseCandidate.ping == nil
- (UPDATE=please-ack): LeaseCandidate.ping is later than LeaseCandidate.renewTime

Would something like this work? Is there possibility of clock drift causing issues?

Reply:
> Unsetting this field will directly trigger a re-election of the corresponding LeaseName that this object references.

If I have 100 LeaseCandidates, nothing that any ONE of them does can trigger an election, right?  The CLE needs to ping _each one_, then wait for some amount of time for _ALL_ of them to ACK (update `renewTime`) or timeout (some candidates may be "dead").  Only after it has an opinion on _every_ candidate can it run the election.  Or am I missing something?

FWIW, leaving `Ping` set is OK, but it feels like detritus that should be cleaned up to me.  But I don't feel strongly.

Reply:
Also, the less we do time-comparisons, the happier I am :)

Reply:
>> If I have 100 LeaseCandidates, nothing that any ONE of them does can trigger an election, right?
>
> A change to the Strategy for instance could cause the current leader to be suboptimal so an election is triggered.

I think this is an open question - does CLE change it's chosen stratgey because ONE candidate said so?  Or because some plurality or majority said so?  

>> The CLE needs to ping each one, then wait for some amount of time for ALL of them to ACK (update renewTime) or timeout (some candidates may be "dead")
>
> Correct. We will wait for whichever one happens first between the timeout and all acks.

So I am a bit confused about what you said earlier: "Unsetting this field will directly trigger a re-election of the corresponding LeaseName that this object references."

For clarity, the way I see the process:
1) Assume a pool of 5 LeaseCandidates [0-4] which all agree on strategy
2) CLE has already made a decision and candidate [3] holds the lease
3) Something happens (async to lease renewal) and the CLE wants to make a different decision
4) CLE pings each candidate via the LeaseCandidate object (* see below)
5) Each candidate sees the ping and ACKs
6) CLE waits for either: a) all 5 candidates to ACK; or b) timeout
7) Of the candidates who ACKed, CLE makes a decision
8) If the decision changes the leader, CLE requests end-of-term via the Lease object
9) CLE waits for either: a) leader ACKs end of term (** see below) or b) lease expires
10) CLE writes new leader
11) New leader acquires the lease

*LeaseCandidate ping: What we need is a clear request and a clear ack.  The request could be as simple as setting `ProofOfLife bool` to `true`, but I sort of hate bool fields.  We could make it a timestamp, so observers could know how long it has been since that ping was sent, and so new pings could be sent even if no response.  We could make it a pointer to a struct, which gives room to expand later (though I am not sure what else we might want to add -- a reason string?).  The response can be simple: set `renewTime`.  The question then is whether we leave the request information or clean it up.

**Lease end-of-term:  Some as above - clear request and ack.  Could be a bool, timestamp, struct, or `PreferredHolder string`.  It all depends on what information we want to be presented to observers and the outgoing lease holder.  Same question about cleanup.

Reply:
> I think this is an open question - does CLE change it's chosen stratgey because ONE candidate said so? Or because some plurality or majority said so?

One candidate, isn't this why we have the superseding logic? If one candidate comes in and has a strategy that supersedes the current strategy, a new leader election should be done.

> 3. Something happens (async to lease renewal) and the CLE wants to make a different decision

This is a bit ambiguous right? A candidate could be just acking and we wouldn't want to trigger an election. From a controller's perspective, it just got an update notification. That's the main reason I'd like to differentiate between different updates.

> *LeaseCandidate ping: What we need is a clear request and a clear ack. The request could be as simple as setting ProofOfLife bool to true, but I sort of hate bool fields. We could make it a timestamp, so observers could know how long it has been since that ping was sent, and so new pings could be sent even if no response. We could make it a pointer to a struct, which gives room to expand later (though I am not sure what else we might want to add -- a reason string?). The response can be simple: set renewTime. The question then is whether we leave the request information or clean it up.

Yes, differentiating between the three updates: clear request, clear ack, clear change of selection strategy/field. I proposed my solution without cleanup so any controller can differentiate between the three types of updates. Obviously, this can also be done with the ProofOfLife being an enum toggle between `request`, `ack` and `reelect` but that's probably much less idiomatic.

> **Lease end-of-term: Some as above - clear request and ack. Could be a bool, timestamp, struct, or PreferredHolder string. It all depends on what information we want to be presented to observers and the outgoing lease holder. Same question about cleanup.

I think we're in agreement on `PreferredHolder` at the moment? This field also targets Lease which is an existing type, so we need to deal with the case that older clients do not understand the annotation, and there's also a clear no-ack case. 

- Request is setting the PreferredHolder field by the CLE controller
- ACK is clearing the HolderIdentity field
- NOACK is continuing to renew despite seeing the PreferredHolder
- Cleanup involves just resetting the PreferredHolder.

Reply:
>> I think this is an open question - does CLE change it's chosen stratgey because ONE candidate said so? Or because some plurality or majority said so?
> 
> One candidate, isn't this why we have the superseding logic? If one candidate comes in and has a strategy that supersedes the current strategy, a new leader election should be done.

OK, so a single candidate can trigger a new election if and only if it has a material change in the election's parameters (in this case the strategy).

>> Something happens (async to lease renewal) and the CLE wants to make a different decision
>
> This is a bit ambiguous right? A candidate could be just acking and we wouldn't want to trigger an election. From a controller's perspective, it just got an update notification. 

The candidate doesn't "want to trigger an election".  They are asserting their configuration, which might trigger an election, or might not.  Don't think about the updates as events with meaning, they are simply carrying state, and the state is unambiguous.

A LeaseCandidate might be updated at any time.  Upon receipt of an update, actors consider the total state of the object.  

> That's the main reason I'd like to differentiate between different updates.

Suppose CLE decides it needs a new election.  It looks at each LC object and finds one whose renewTime is 1 second ago -- does it really need a ping?  My point is that the updates aren't different, they both indicate "this actor is alive as of this timestamp".

> Yes, differentiating between the three updates: clear request, clear ack, clear change of selection strategy/field. I proposed my solution without cleanup so any controller can differentiate between the three types of updates.

Describe them instead as states.

1) Steady state: `renewTime` is set, `ping` is not set, strategy is whatever value
    a) maybe dead: `renewTime` is older than X duration
    b) presumed alive: `renewTime` is less than X duration
2) Proof-of-life requested: `renewTime` is set, `ping` is not set, strategy is whatever value

There's only those 2 (or 3 if you split up case 1) states for a LeaseCandidate.  There is no state which is "requesting a strategy change".  CLE can look at the LeaseCandidates requested strategies and the Lease's current strategy and recognize that a change in the active strategy is needed, but no one candidate needs to express that as a state.

Reply:
Okay.

- Steady state: renewTime is set, ping is not set, strategy is whatever value
a) dead: renewTime is older than CandidateDurationSeconds old
b) presumed alive: renewTime is less than CandidateDurationSeconds old
- proof of life requested: renewTime is set, ping is set, strategy is whatever vale

That's is a good point that CLE should determine the strategy change rather than the component. I was hoping the CLE could be less "smart" and just react to the messages in the update, but I guess it makes sense that it should make the decision. I'll keep the API as you described, but still want to go through the example again and figure out the CLE logic of figuring out whether a state requires any changes.

1. Assume a pool of 5 LeaseCandidates [0-4] which all agree on strategy
2. CLE has already made a decision and candidate [3] holds the lease
3. Something happens (async to lease renewal) and the CLE wants to make a different decision
    **- CLE saw an update, compared the LeaseCandidate to the current leader**
5. CLE pings each candidate via the LeaseCandidate object (* see below)
6. Each candidate sees the ping and ACKs
   **- CLE is also watching the candidates, but it can treat pings as noops since it seems that ping is set**
8. CLE waits for either: a) all 5 candidates to ACK; or b) timeout
    **- How does CLE differentiate between an ACK and a config/strategy change? Would it have to compare an ACK'd leader to the current best leader every time? Would it be based on the sole fact that an async leader election is already going on?**
9. Of the candidates who ACKed, CLE makes a decision
10. If the decision changes the leader, CLE requests end-of-term via the Lease object
11. CLE waits for lease expiry or holderIdentity being unset
12. CLE writes new leader
13. New leader acquires the lease

Reply:
I'll break my reply into 2 parts.

First, I want to repeat (and expand on) a question that was buried in my comment:

Suppose CLE decides it needs a new election. It looks at each LC object and finds one whose renewTime is 1 second ago -- does it really need to ping that LC?  Or is 1 second "proof enough"?  What if it was 1 millisecond old?

I want to emphasize two things about writing controllers which Kubernetes highlights:
 1) There's no such thing as "instantaneous" so any comparison of timestamps has to carry some "acceptable window".
 2) Any process can crash and restart at ANY POINT in the code, so we must write the code to behave correct in the worst case.  This is why state matters more than events.

What I am getting at is that I think there is a useful state you are not considering, which is "recently renewed".  Consider this case:

* A LeaseCandidate joins the pool with an older compat version
* CLE decides that the new guy should be in charge
* CLE writes the "proof of life" request to the first LC
* apiserver crashes!  oh no!
* The pinged LC is ACKed
* apiserver comes back up in a second or two and restarts CLE controller
* CLE looks at all leases with strategy != "", finds this lease 
* CLE looks at all lease candidates for this lease and decides which one SHOULD be the leader
* CLE compares this with the lease's holder, and finds they do not match

NOTE: CLE "forgot" that it made this decision and that it already sent some pings, and it missed the LC's ACK event.   But it can see the state (`renewTime` is 2 seconds ago, `ping` is nil).  Should it re-ping?  Or is it OK to use the result of the last ping?  

I think there's some time window where it MUST be OK.  If it was updated 5 minutes ago, that's pretty clearly not OK.  If it was updated 1 millisecond ago, that's pretty clearly OK.  So there's an edge somewhere in there.  3 seconds?  5 seconds?  The lease's duration (e.g. 15sec)?

Reply:
> still want to go through the example again and figure out the CLE logic of figuring out whether a state requires any changes.

1) Assume a pool of 5 LeaseCandidates [0-4] which all agree on strategy
2) CLE has already made a decision and candidate [3] holds the lease
3) Something happens (async to lease renewal) and the CLE wants to make a different decision
  - e.g. CLE saw an LC update which asserts a new preferred strategy
  - e.g. CLE saw a new LC with an olde compatVersion
  - e.g. CLE saw the lease expire without being renewed
4) CLE pings each not-recently-renewed candidate via their LeaseCandidate objects
  - CLE will get events for those LC updates, but the new state is neither renewed nor timed out
5) Each living candidate sees its ping and eventually ACKs (set renewTime)
6) CLE waits for all candidates to either a) ACK; or b) timeout
  - If CLE observes a config/strategy change it can either abort this election and go back to (3) or it can finish this election and catch it on the next loop (everything is async, so it's a race no matter what)
7) Of the candidates who ACKed, CLE makes a decision
8) If the decision changes the leader, CLE requests end-of-term via the Lease object
9) CLE waits for either a) the lease to expire; or b) the holderIdentity to be unset (reliquish)
  - if the lease holder does not reliquish, CLE needs to do something - events, logs, conditions, etc
10) CLE writes the new leader to the lease
11) The new leader acquires the lease and sets `acquireTime` and `renewTime`

Reply:
> I think there's some time window where it MUST be OK. If it was updated 5 minutes ago, that's pretty clearly not OK. If it was updated 1 millisecond ago, that's pretty clearly OK. So there's an edge somewhere in there. 3 seconds? 5 seconds? The lease's duration (e.g. 15sec)?

Yes, spec.CandidateDurationSeconds is used for this. I have this set at 5s for our controllers.

Reply:
I thought candidate durations were long, like O(minutes) ?

Reply:
ah I see where this needs clarification. duration is intended to be the length that the candidate is valid for without needing an additional ping for reelection. When a lease reelection is done, it triggers on all candidates that renewed earlier than lastRenewTime + duration.

We were also talking about a separate definition of duration as how long ago a candidate last renewed before it's considered "dead" and out of scope for re-pinging. I think if a candidate hits some really long interval like this it should just be gc'd (deleted) instead of remaining active in a dormant state.

Reply:
What's the use-case for a candidate whose duration is longer than the lease's?  In fact, why does candidate need a duration at all?

Reply:
I just thought it'd be a nice thing to define client side. Hm that's a good point. If a lease holder goes away and candidateDuration > leaseDuration, it's possible to elect the old candidate which has gone away already. So candidateDuration should be <= leaseDuration. Technically not necessary if we can agree on a common duration for all components.

Reply:
I see two needs:

1) candidate is fresh enough to contend in an election: something like `time.Since(candidate.renewTime) < lease.duration`
2) candidate should be garbage-collected: is this what candidate duration is?

Consider this case:
* Deploy a component with 3 replicas
* 3 LCs [0-2]
* LC[1] is elected
* hours, days, weeks pass, LC[1] is still the leader

Did anyone update the other LCs?  How often?  Where is that defined?

* The component is turned off, but doesn't clean up its LCs

Do they get auto-cleaned up, or do they linger forever?  What signal does GC use to know they are safe to clean up?  Is it "we are GC'ing the Lease, so we can GC the candidates, too" ?

Reply:
Just make sure that is documented!